File: src/class034/C1_IntersectionOfTwoLinkedLists.py
def getIntersectionNode(h1, h2):  
    if not h1 or not h2:  
        return None  
    a, b = h1, h2  #a就是长链表
    diff = 0  
    while a.next:  
        a = a.next  
        diff += 1  
    while b.next:  
        b = b.next  
        diff -= 1  
    if a != b:  
        return None  
    # 👆上方确认是否有相交的节点👆 + diff
    

    # 赋值长短链头部
    if diff >= 0:  
        a, b = h1, h2  
    else:  
        a, b = h2, h1  



    #长链表先走差值步，ab相遇点即相交位点+双while
    diff = abs(diff)  
    while diff:  #长链表先走差值步骤
        a = a.next  
        diff -= 1  
    while a != b:  # ab相遇点，跳出while loop，返回相交的位置
        a = a.next  
        b = b.next  
    return a  

File: src/class034/C1_IntersectionOfTwoLinkedLists.cpp
#include <iostream>

using namespace std;

// 返回两个无环链表相交的第一个节点
// 测试链接 : https://leetcode-cn/problems/intersection-of-two-linked-lists/
struct ListNode {
    int val;
    ListNode *next;
};

ListNode *getIntersectionNode(ListNode *h1, ListNode *h2) {
    if (h1 == nullptr || h2 == nullptr) {
        return nullptr;
    }
    ListNode *a = h1, *b = h2;
    int diff = 0;
    while (a->next != nullptr) {
        a = a->next;
        diff++;
    }
    while (b->next != nullptr) {
        b = b->next;
        diff--;
    }
    if (a != b) {
        return nullptr;
    }
    if (diff >= 0) {
        a = h1;
        b = h2;
    } else {
        a = h2;
        b = h1;
    }
    diff = abs(diff);
    while (diff-- != 0) {
        a = a->next;
    }
    while (a != b) {
        a = a->next;
        b = b->next;
    }
    return a;
}

// For testing purpose, not to be submitted.
int main() {
    // Test cases can be added here.
    return 0;
}
